doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+
+Now let's talk about what we're going to do.
+Our mission (and we have no choice but to accept it) is to obfuscate all null literals in the code.
+It means that we are going to replace (almost) all the zeroes in the code by a non-trivial boolean expression always false.
+
+        prime1 * ((x | any1)**2) != prime2 * ((y | any2)**2)
+
+    - prime1 and prime2 are *distinct* prime numbers
+    - any1 and any2 are *distinct* strictly positive random numbers
+    - x and y are two variables picked from the program (they have to be reachable from the obfuscation instructions)
+
+This expression will always return a boolean zero (false). The idea is to insert this test into our code, just before the 0 we want to obfuscate and to replace this 0 by the result of our comparison.
+As you have probably noticed we will have to pay attention to the type of the original 0 and make sure we cast the result of our expression to its type.
+
+This obfuscation may not be the most sophisticated ever written but it's enough to learn the basics of LLVM bytecode obfuscation (and to annoy our reverse engineer friends.)
+
+Requirements
+============
+
+Programming background
+**********************
+To go through this tutorial you only need to be able to read C/C++ code. We will learn the basics of LLVM API and LLVM bytecode together.
+
+Environment
+***********
+If you want to experiment along with this tutorial (which is strongly recommended) you will need to set up an LLVM development environment.
+
+    * First let's download the LLVM sources...
+        ``>$ git clone --depth=1 --branch=release_35 https://github.com/llvm-mirror/llvm llvm_src``
+
+    * Now download the clang sources inside the llvm_src/tools directory.
+        ``>$ cd llvm_src/tools && git clone --depth=1 --branch=release_35 https://github.com/llvm-mirror/clang``
+
+    * Create a build directory somewhere *out of the llvm source tree*:
+        ``>$ mkdir path/to/llvm/build/llvm``
+
+    * Let's build LLVM:
+        ``>$ cd ~/llvm && cmake path/to/llvm/sources``
+        ``>$ make -j``
+
+    * And... wait....
+        ``>IRL sleep 1000``
+
+    * (Optional) You should set your path to include the freshly baked LLVM tools and clang:
+        ``>$ export PATH=~path/to/llvm/build/llvm/bin:$PATH``
+
+    * (Optional) You can run the test suite to make sure that LLVM was built correctly:
+        ``>$ make check``
+
+You have just built LLVM and clang but we going to build the passes out of the LLVM source tree. To do so we have prepared a git repository with the basic infrastructure.
+
+        ``>$ git clone https://github.com/quarkslab/llvm-passes``
+
+From now on we will be working exclusively inside the ``llvm-passes`` folder (we will refer to it as ``$PASSDIR``). So let's visit our new office:
+    * *cmake*: cmake definitions to check the python environment. Required to generate our passes test suites.
```

Comment:
to link to the LLVM environment and also check the Python environment for a few extra tools (we are using `lit (ref)` to run the validation. Wait. Did I wrote validation?)